go/consensus/tendermint/abci/messages.go

```diff
@@ -21,16 +21,20 @@ func (md *messageDispatcher) Subscribe(kind interface{}, ms api.MessageSubscribe
 }
 
 // Implements api.MessageDispatcher.
-func (md *messageDispatcher) Publish(ctx *api.Context, kind, msg interface{}) error {
-	if len(md.subscriptions[kind]) == 0 {
-		return api.ErrNoSubscribers
+func (md *messageDispatcher) Publish(ctx *api.Context, kind, msg interface{}) ([]interface{}, error) {
```

Comment:
now that I actually see the tests, the interface for this feels so idiosyncratic. for one, it's one of those rare functions where the first return value is meaningful when the error is non-nil.

from the code in this PR, the behavior is like this:

- the results slice always has one element per subscriber
- the error is a `multierror` with an element for each subscriber that errored, thus varying length
- the subscriber's interface treats its response as undefined when the error is non-nil
- for a subscriber that errors, its corresponding slot in the results is left as nil

I just wonder if we will really be able to write rigorous multi-subscriber programs with this. my concerns are:

1. it's not clear to what extent the code calling Publish is expected to know the composition and ordering of the subscribers. the previous code where the errors are piled together suggest the calling code doesn't know/doesn't care. but surely that knowledge must be increased now that we have responses that the calling code will be looking for.
2. it's not clear how the calling code should find the response it's looking for. by known index? or should it maybe loop through and find it by type?
3. some calling code that's less knowledgeable about the composition of subscribers won't be able to differentiate subscribers that succeed and return `nil` from subscribers that errored. at least not from looking at the results slice. not sure if this will come up.
4. for calling code that's more knowledgeable about the composition of subscribers, I think it will be hard to match up errors with what subscriber it came from. maybe those use cases will have to include more metadata in the errors to work around that.

do we have cases where there are multiple subscribers for a message? or if not, what are the intended use cases? how do/will we handle the situation where some subscribers fail and some succeed? is the zeroth result always the one that a runtime will want to see?

Reply:
> the error is a multierror with an element for each subscriber that errored, thus varying length

The error return type should probably be changed to be a list of errors, so then the caller knows which responses are non failed.

> do we have cases where there are multiple subscribers for a message? or if not, what are the intended use cases? 

No, no such cases. Not sure about intended use-cases.

> how do/will we handle the situation where some subscribers fail and some succeed

I guess it's up to the caller to decide.

> is the zeroth result always the one that a runtime will want to see?

For all existing cases, yes, as there's always a single subscriber. Maybe the could should propagate all results to the caller, and not just the first one (although this makes no difference at this point in time).

Reply:
There's at least a use case for zero subscribers, so that we can support non-lock-step changes to paratimes and core. Maybe now is the time to consider switching to a zero-or-one subscriber model of message handlers?

Reply:
error slice would be unambiguous I guess. potentially verbose retrofitting work though.

I see some places in this PR's callsite changes like this

```go
// Notify other interested applications about the resumed runtime.
if _, err = app.md.Publish(ctx, registryApi.MessageRuntimeResumed, rt); err != nil {
```

I guess we've been using this message dispatcher class for things other than runtime-sent messages too

Reply:
> I guess we've been using this message dispatcher class for things other than runtime-sent messages too.

Yes, this is a general pubsub mechanism used by the consensus layer services.

Reply:
Ok agreed, i'll make `Publish` return a single reuslt then

Reply:
Updated. Please take a look, I'll update the ADR once this new approach is finalized.